contracts/schemes/JoinAndQuit.sol

```diff
@@ -0,0 +1,243 @@
+            _descriptionHash,
+            member,
+            _fundAmount
+        );
+
+        proposalsInfo[address(votingMachine)][proposalId] = ProposalInfo({
+            blockNumber:block.number,
+            avatar:avatar
+        });
+        return proposalId;
+    }
+
+    /**
+    * @dev RedeemReputation reward for proposal
+    * @param _proposalId the ID of the voting in the voting machine
+    * @return reputation the redeemed reputation.
+    */
+    function redeemReputation(bytes32 _proposalId) public returns(uint256 reputation) {
+
+        Proposal memory _proposal = proposals[_proposalId];
+        Proposal storage proposal = proposals[_proposalId];
+        //set proposal proposedMember to zero to prevent reentrancy attack.
+        proposal.proposedMember = address(0);
+        require(proposal.accepted == true, " proposal not accepted");
+        require(
+        Controller(
+        avatar.owner()).mintReputation(memberReputation, _proposal.proposedMember));
+        proposal.proposedMember = _proposal.proposedMember;
+        emit RedeemReputation(address(avatar), _proposalId, _proposal.proposedMember, memberReputation);
+    }
+
+    /**
+    * @dev donate donate to the dao. no repution is minted
+    */
+    function donate(uint256 _donation) public {
+        address(fundingToken).safeTransferFrom(msg.sender, address(avatar), _donation);
+        fundings[msg.sender] = fundings[msg.sender].add(_donation);
+        totalDonation = totalDonation.add(_donation);
+        checkFundedBeforeDeadLine();
+        emit Donation(address(avatar), _donation);
+    }
+
+    /**
+    * @dev rageQuit quit from the dao.
+    * can be done on any time
+    * REFUND = USER_DONATION * CURRENT_DAO_BALANCE / TOTAL_DONATIONS
+    * @return refund the refund amount
+    */
+    function rageQuit() public returns(uint256 refund) {
+        require(fundings[msg.sender] > 0, "no fund to RageQuit");
+        uint256 userDonation = fundings[msg.sender];
+        fundings[msg.sender] = 0;
+        refund = userDonation.mul(fundingToken.balanceOf(address(avatar))).div(totalDonation);
+        totalDonation = totalDonation.sub(userDonation);
+        require(
+        Controller(
+        avatar.owner()).externalTokenTransfer(fundingToken, msg.sender, refund));
+        emit RageQuit(address(avatar), refund);
+    }
+
+    /**
+    * @dev checkFundedBeforeDeadLine check if funding goal reached.
+    */
+    function checkFundedBeforeDeadLine() private {
+        if ((avatar.db(FUNDED_BEFORE_DEADLINE_KEY).hashCompareWithLengthCheck(FUNDED_BEFORE_DEADLINE_VALUE) == false) &&
+            (fundingToken.balanceOf(address(avatar)) >= fundingGoal) &&
+            // solhint-disable-next-line not-rely-on-time
+            (now < fundingGoalDeadLine)) {
+            require(
+            Controller(
+            avatar.owner()).setDBValue("FUNDED_BEFORE_DEADLINE", "TRUE"));
+            emit FundedDeadLineReached(address(avatar));
```

Comment:
This may very well happen way before the funding deadline is reached, so this event name is confusing.

`DAOIsFunded` would be better, or even `FundedBeforeDeadline` to be consistent with the db key.